File: app/src/main/java/com/bscan/model/graph/entities/ScanDataEntities.kt
package com.bscan.model.graph.entities

import com.bscan.model.graph.Entity
import com.bscan.model.graph.PropertyValue
import com.bscan.model.graph.ValidationResult
import java.time.LocalDateTime

/**
 * Scan occurrence entity - represents each individual scan event
 * PERSISTENT: One created for each scan
 */
open class ScanOccurrence(
    id: String = generateId(),
    label: String,
    properties: MutableMap<String, PropertyValue> = mutableMapOf()
) : Activity(
    id = id,
    activityType = ActivityTypes.SCAN,
    label = label,
    properties = properties
) {
    
    var deviceInfo: String?
        get() = getProperty("deviceInfo")
        set(value) { value?.let { setProperty("deviceInfo", it) } }
    
    var scanLocation: String?
        get() = getProperty("scanLocation") 
        set(value) { value?.let { setProperty("scanLocation", it) } }
    
    var scanMethod: String?
        get() = getProperty("scanMethod")  // "nfc", "qr", "barcode", "manual"
        set(value) { value?.let { setProperty("scanMethod", it) } }
    
    var appVersion: String?
        get() = getProperty("appVersion")
        set(value) { value?.let { setProperty("appVersion", it) } }
    
    var userData: String?
        get() = getProperty("userData")  // User annotations, notes
        set(value) { value?.let { setProperty("userData", it) } }
}

/**
 * Raw scan data entity - stores unique encoded/encrypted data
 * PERSISTENT: Shared across multiple scan occurrences via relationships
 * Deduplicated based on content hash
 */
open class RawScanData(
    id: String = generateId(),
    label: String,
    val scanFormat: String,  // "bambu_rfid", "creality_rfid", "qr_code", "barcode"
    properties: MutableMap<String, PropertyValue> = mutableMapOf()
) : Information(
    id = id,
    informationType = "raw_scan_data",
    label = label,
    properties = properties
) {
    
    init {
        setProperty("scanFormat", scanFormat)
    }
    
    var rawData: String
        get() = getProperty("rawData") ?: ""
        set(value) { setProperty("rawData", value) }
    
    var dataSize: Int
        get() = getProperty("dataSize") ?: 0
        set(value) { setProperty("dataSize", value) }
    
    var contentHash: String?
        get() = getProperty("contentHash")
        set(value) { value?.let { setProperty("contentHash", it) } }
    
    var encoding: String?
        get() = getProperty("encoding")  // "hex", "base64", "utf8"
        set(value) { value?.let { setProperty("encoding", it) } }
    
    var checksumValid: Boolean?
        get() = getProperty("checksumValid")
        set(value) { value?.let { setProperty("checksumValid", it) } }
    
    override fun copy(newId: String): RawScanData {
        return RawScanData(
            id = newId,
            label = label,
            scanFormat = scanFormat,
            properties = properties.toMutableMap()
        )
    }
    
    override fun validate(): ValidationResult {
        val errors = mutableListOf<String>()
        
        if (scanFormat.isBlank()) errors.add("Scan format must be specified")
        if (rawData.isBlank()) errors.add("Raw data cannot be empty")
        
        return if (errors.isEmpty()) {
            ValidationResult.valid()
        } else {
            ValidationResult.invalid(*errors.toTypedArray())
        }
    }
}

/**
 * Decoded encrypted entity - non-encrypted metadata from raw scan
 * EPHEMERAL: Generated on-demand with TTL caching
 */
open class DecodedEncrypted(
    id: String = generateId(),
    label: String,
    properties: MutableMap<String, PropertyValue> = mutableMapOf()
) : Information(
    id = id,
    informationType = "decoded_encrypted",
    label = label,
    properties = properties
) {
    
    var tagType: String?
        get() = getProperty("tagType")
        set(value) { value?.let { setProperty("tagType", it) } }
    
    var tagUid: String?
        get() = getProperty("tagUid")
        set(value) { value?.let { setProperty("tagUid", it) } }
    
    var dataBlocks: Int?
        get() = getProperty("dataBlocks")
        set(value) { value?.let { setProperty("dataBlocks", it) } }
    
    var sectorCount: Int?
        get() = getProperty("sectorCount")
        set(value) { value?.let { setProperty("sectorCount", it) } }
    
    var authenticated: Boolean?
        get() = getProperty("authenticated")
        set(value) { value?.let { setProperty("authenticated", it) } }
    
    var keyDerivationTime: Long?
        get() = getProperty<Long>("keyDerivationTime")
        set(value) { value?.let { setProperty("keyDerivationTime", it) } }
    
    var cacheTimestamp: LocalDateTime
        get() = getProperty("cacheTimestamp") ?: LocalDateTime.now()
        set(value) { setProperty("cacheTimestamp", value) }
    
    var cacheTtlMinutes: Int
        get() = getProperty("cacheTtlMinutes") ?: 60
        set(value) { setProperty("cacheTtlMinutes", value) }
    
    fun isExpired(): Boolean {
        return LocalDateTime.now().isAfter(cacheTimestamp.plusMinutes(cacheTtlMinutes.toLong()))
    }
}

/**
 * Encoded decrypted entity - decrypted but still hex-encoded data  
 * EPHEMERAL: Generated on-demand with TTL caching
 */
open class EncodedDecrypted(
    id: String = generateId(),
    label: String,
    properties: MutableMap<String, PropertyValue> = mutableMapOf()
) : Information(
    id = id,
    informationType = "encoded_decrypted", 
    label = label,
    properties = properties
) {
    
    var decryptedData: String
        get() = getProperty("decryptedData") ?: ""
        set(value) { setProperty("decryptedData", value) }
    
    var blockStructure: String?
        get() = getProperty("blockStructure")  // JSON representation of block layout
        set(value) { value?.let { setProperty("blockStructure", it) } }
    
    var keyInfo: String?
        get() = getProperty("keyInfo")  // Which keys were used for decryption
        set(value) { value?.let { setProperty("keyInfo", it) } }
    
    var decryptionTime: Long?
        get() = getProperty<Long>("decryptionTime")
        set(value) { value?.let { setProperty("decryptionTime", it) } }
    
    var cacheTimestamp: LocalDateTime
        get() = getProperty("cacheTimestamp") ?: LocalDateTime.now()
        set(value) { setProperty("cacheTimestamp", value) }
    
    var cacheTtlMinutes: Int
        get() = getProperty("cacheTtlMinutes") ?: 30
        set(value) { setProperty("cacheTtlMinutes", value) }
    
    fun isExpired(): Boolean {
        return LocalDateTime.now().isAfter(cacheTimestamp.plusMinutes(cacheTtlMinutes.toLong()))
    }
}

/**
 * Decoded decrypted entity - fully interpreted structured data
 * EPHEMERAL: Generated on-demand with TTL caching
 */
open class DecodedDecrypted(
    id: String = generateId(),
    label: String,
    properties: MutableMap<String, PropertyValue> = mutableMapOf()
) : Information(
    id = id,
    informationType = "decoded_decrypted",
    label = label,
    properties = properties
) {
    
    var interpretedData: String?
        get() = getProperty("interpretedData")  // JSON representation of structured data
        set(value) { value?.let { setProperty("interpretedData", it) } }
    
    var filamentProperties: String?
        get() = getProperty("filamentProperties")  // Bambu-specific: material, color, etc.
        set(value) { value?.let { setProperty("filamentProperties", it) } }
    
    var productInfo: String?
        get() = getProperty("productInfo")  // SKU, manufacturer, model
        set(value) { value?.let { setProperty("productInfo", it) } }
    
    var temperatureSettings: String?
        get() = getProperty("temperatureSettings")  // Print/bed temperatures
        set(value) { value?.let { setProperty("temperatureSettings", it) } }
    
    var physicalProperties: String?
        get() = getProperty("physicalProperties")  // Mass, dimensions, etc.
        set(value) { value?.let { setProperty("physicalProperties", it) } }
    
    var identifiers: String?
        get() = getProperty("identifiers")  // UID, tray ID, consumable ID
        set(value) { value?.let { setProperty("identifiers", it) } }
    
    var interpretationVersion: String?
        get() = getProperty("interpretationVersion")  // Version of interpretation logic
        set(value) { value?.let { setProperty("interpretationVersion", it) } }
    
    var interpretationTime: Long?
        get() = getProperty<Long>("interpretationTime")
        set(value) { value?.let { setProperty("interpretationTime", it) } }
    
    var cacheTimestamp: LocalDateTime
        get() = getProperty("cacheTimestamp") ?: LocalDateTime.now()
        set(value) { setProperty("cacheTimestamp", value) }
    
    var cacheTtlMinutes: Int
        get() = getProperty("cacheTtlMinutes") ?: 15  // Shorter TTL for complex interpretations
        set(value) { setProperty("cacheTtlMinutes", value) }
    
    fun isExpired(): Boolean {
        return LocalDateTime.now().isAfter(cacheTimestamp.plusMinutes(cacheTtlMinutes.toLong()))
    }
}

/**
 * Common scan data relationship types
 */
object ScanDataRelationshipTypes {
    const val SCANNED = "scanned"                    // ScanOccurrence -> RawScanData
    const val DECODED_TO = "decoded_to"              // RawScanData -> DecodedEncrypted
    const val DECRYPTED_TO = "decrypted_to"          // RawScanData -> EncodedDecrypted
    const val INTERPRETED_AS = "interpreted_as"      // RawScanData -> DecodedDecrypted
    const val DERIVED_FROM = "derived_from"          // Any ephemeral -> source entity
    const val CREATED_ENTITIES = "created_entities"  // DecodedDecrypted -> PhysicalComponent entities
}

/**
 * Cache management for ephemeral entities
 */
data class CacheEntry<T : Entity>(
    val entity: T,
    val timestamp: LocalDateTime = LocalDateTime.now(),
    val ttlMinutes: Int
) {
    fun isExpired(): Boolean {
        return LocalDateTime.now().isAfter(timestamp.plusMinutes(ttlMinutes.toLong()))
    }
    
    fun getRemainingTtl(): Long {
        val expiryTime = timestamp.plusMinutes(ttlMinutes.toLong())
        val now = LocalDateTime.now()
        return if (now.isBefore(expiryTime)) {
            java.time.Duration.between(now, expiryTime).toMinutes()
        } else {
            0
        }
    }
}

/**
 * Cache statistics for monitoring
 */
data class CacheStatistics(
    val totalEntries: Int,
    val expiredEntries: Int,
    val hitRate: Float,
    val memoryUsageBytes: Long,
    val oldestEntryAge: Long,
    val averageTtl: Int
)

File: src/models/entities/ScanDataEntities.ts
/**
 * Scan data entities for RFID/NFC/QR processing pipeline
 */

import { Activity, Information } from './CoreEntities';
import { Entity, ValidationResult, generateId } from './Entity';
import { PropertyValue } from './PropertyValue';
import { ActivityTypes, CacheEntry } from './types';

/**
 * Scan occurrence entity - represents each individual scan event
 * PERSISTENT: One created for each scan
 */
export class ScanOccurrence extends Activity {
  constructor(
    id: string = generateId(),
    label: string,
    properties: Map<string, PropertyValue> = new Map()
  ) {
    super(id, ActivityTypes.SCAN, label, properties);
  }

  get deviceInfo(): string | undefined {
    return this.getProperty<string>('deviceInfo');
  }
  
  set deviceInfo(value: string | undefined) {
    if (value !== undefined) this.setProperty('deviceInfo', value);
  }

  get scanLocation(): string | undefined {
    return this.getProperty<string>('scanLocation');
  }
  
  set scanLocation(value: string | undefined) {
    if (value !== undefined) this.setProperty('scanLocation', value);
  }

  get scanMethod(): string | undefined {
    return this.getProperty<string>('scanMethod'); // "nfc", "qr", "barcode", "manual"
  }
  
  set scanMethod(value: string | undefined) {
    if (value !== undefined) this.setProperty('scanMethod', value);
  }

  get appVersion(): string | undefined {
    return this.getProperty<string>('appVersion');
  }
  
  set appVersion(value: string | undefined) {
    if (value !== undefined) this.setProperty('appVersion', value);
  }

  get userData(): string | undefined {
    return this.getProperty<string>('userData'); // User annotations, notes
  }
  
  set userData(value: string | undefined) {
    if (value !== undefined) this.setProperty('userData', value);
  }
}

/**
 * Raw scan data entity - stores unique encoded/encrypted data
 * PERSISTENT: Shared across multiple scan occurrences via relationships
 * Deduplicated based on content hash
 */
export class RawScanData extends Information {
  constructor(
    id: string = generateId(),
    label: string,
    public readonly scanFormat: string, // "bambu_rfid", "creality_rfid", "qr_code", "barcode"
    properties: Map<string, PropertyValue> = new Map()
  ) {
    super(id, 'raw_scan_data', label, properties);
    this.setProperty('scanFormat', scanFormat);
  }

  get rawData(): string {
    return this.getProperty<string>('rawData') ?? '';
  }
  
  set rawData(value: string) {
    this.setProperty('rawData', value);
  }

  get dataSize(): number {
    return this.getProperty<number>('dataSize') ?? 0;
  }
  
  set dataSize(value: number) {
    this.setProperty('dataSize', value);
  }

  get contentHash(): string | undefined {
    return this.getProperty<string>('contentHash');
  }
  
  set contentHash(value: string | undefined) {
    if (value !== undefined) this.setProperty('contentHash', value);
  }

  get encoding(): string | undefined {
    return this.getProperty<string>('encoding'); // "hex", "base64", "utf8"
  }
  
  set encoding(value: string | undefined) {
    if (value !== undefined) this.setProperty('encoding', value);
  }

  get checksumValid(): boolean | undefined {
    return this.getProperty<boolean>('checksumValid');
  }
  
  set checksumValid(value: boolean | undefined) {
    if (value !== undefined) this.setProperty('checksumValid', value);
  }

  copy(newId: string = generateId()): RawScanData {
    return new RawScanData(
      newId,
      this.label,
      this.scanFormat,
      new Map(this.properties)
    );
  }

  validate(): ValidationResult {
    const errors: string[] = [];
    
    if (!this.scanFormat.trim()) {
      errors.push('Scan format must be specified');
    }
    
    if (!this.rawData.trim()) {
      errors.push('Raw data cannot be empty');
    }
    
    return errors.length === 0 
      ? ValidationResult.valid() 
      : ValidationResult.invalid(...errors);
  }
}

/**
 * Decoded encrypted entity - non-encrypted metadata from raw scan
 * EPHEMERAL: Generated on-demand with TTL caching
 */
export class DecodedEncrypted extends Information {
  constructor(
    id: string = generateId(),
    label: string,
    properties: Map<string, PropertyValue> = new Map()
  ) {
    super(id, 'decoded_encrypted', label, properties);
  }

  get tagType(): string | undefined {
    return this.getProperty<string>('tagType');
  }
  
  set tagType(value: string | undefined) {
    if (value !== undefined) this.setProperty('tagType', value);
  }

  get tagUid(): string | undefined {
    return this.getProperty<string>('tagUid');
  }
  
  set tagUid(value: string | undefined) {
    if (value !== undefined) this.setProperty('tagUid', value);
  }

  get dataBlocks(): number | undefined {
    return this.getProperty<number>('dataBlocks');
  }
  
  set dataBlocks(value: number | undefined) {
    if (value !== undefined) this.setProperty('dataBlocks', value);
  }

  get sectorCount(): number | undefined {
    return this.getProperty<number>('sectorCount');
  }
  
  set sectorCount(value: number | undefined) {
    if (value !== undefined) this.setProperty('sectorCount', value);
  }

  get authenticated(): boolean | undefined {
    return this.getProperty<boolean>('authenticated');
  }
  
  set authenticated(value: boolean | undefined) {
    if (value !== undefined) this.setProperty('authenticated', value);
  }

  get keyDerivationTime(): number | undefined {
    return this.getProperty<number>('keyDerivationTime');
  }
  
  set keyDerivationTime(value: number | undefined) {
    if (value !== undefined) this.setProperty('keyDerivationTime', value);
  }

  get cacheTimestamp(): Date {
    return this.getProperty<Date>('cacheTimestamp') ?? new Date();
  }
  
  set cacheTimestamp(value: Date) {
    this.setProperty('cacheTimestamp', value);
  }

  get cacheTtlMinutes(): number {
    return this.getProperty<number>('cacheTtlMinutes') ?? 60;
  }
  
  set cacheTtlMinutes(value: number) {
    this.setProperty('cacheTtlMinutes', value);
  }

  isExpired(): boolean {
    const expiryTime = new Date(this.cacheTimestamp.getTime() + this.cacheTtlMinutes * 60 * 1000);
    return new Date() > expiryTime;
  }
}

/**
 * Encoded decrypted entity - decrypted but still hex-encoded data  
 * EPHEMERAL: Generated on-demand with TTL caching
 */
export class EncodedDecrypted extends Information {
  constructor(
    id: string = generateId(),
    label: string,
    properties: Map<string, PropertyValue> = new Map()
  ) {
    super(id, 'encoded_decrypted', label, properties);
  }

  get decryptedData(): string {
    return this.getProperty<string>('decryptedData') ?? '';
  }
  
  set decryptedData(value: string) {
    this.setProperty('decryptedData', value);
  }

  get blockStructure(): string | undefined {
    return this.getProperty<string>('blockStructure'); // JSON representation of block layout
  }
  
  set blockStructure(value: string | undefined) {
    if (value !== undefined) this.setProperty('blockStructure', value);
  }

  get keyInfo(): string | undefined {
    return this.getProperty<string>('keyInfo'); // Which keys were used for decryption
  }
  
  set keyInfo(value: string | undefined) {
    if (value !== undefined) this.setProperty('keyInfo', value);
  }

  get decryptionTime(): number | undefined {
    return this.getProperty<number>('decryptionTime');
  }
  
  set decryptionTime(value: number | undefined) {
    if (value !== undefined) this.setProperty('decryptionTime', value);
  }

  get cacheTimestamp(): Date {
    return this.getProperty<Date>('cacheTimestamp') ?? new Date();
  }
  
  set cacheTimestamp(value: Date) {
    this.setProperty('cacheTimestamp', value);
  }

  get cacheTtlMinutes(): number {
    return this.getProperty<number>('cacheTtlMinutes') ?? 30;
  }
  
  set cacheTtlMinutes(value: number) {
    this.setProperty('cacheTtlMinutes', value);
  }

  isExpired(): boolean {
    const expiryTime = new Date(this.cacheTimestamp.getTime() + this.cacheTtlMinutes * 60 * 1000);
    return new Date() > expiryTime;
  }
}

/**
 * Decoded decrypted entity - fully interpreted structured data
 * EPHEMERAL: Generated on-demand with TTL caching
 */
export class DecodedDecrypted extends Information {
  constructor(
    id: string = generateId(),
    label: string,
    properties: Map<string, PropertyValue> = new Map()
  ) {
    super(id, 'decoded_decrypted', label, properties);
  }

  get interpretedData(): string | undefined {
    return this.getProperty<string>('interpretedData'); // JSON representation of structured data
  }
  
  set interpretedData(value: string | undefined) {
    if (value !== undefined) this.setProperty('interpretedData', value);
  }

  get filamentProperties(): string | undefined {
    return this.getProperty<string>('filamentProperties'); // Bambu-specific: material, color, etc.
  }
  
  set filamentProperties(value: string | undefined) {
    if (value !== undefined) this.setProperty('filamentProperties', value);
  }

  get productInfo(): string | undefined {
    return this.getProperty<string>('productInfo'); // SKU, manufacturer, model
  }
  
  set productInfo(value: string | undefined) {
    if (value !== undefined) this.setProperty('productInfo', value);
  }

  get temperatureSettings(): string | undefined {
    return this.getProperty<string>('temperatureSettings'); // Print/bed temperatures
  }
  
  set temperatureSettings(value: string | undefined) {
    if (value !== undefined) this.setProperty('temperatureSettings', value);
  }

  get physicalProperties(): string | undefined {
    return this.getProperty<string>('physicalProperties'); // Mass, dimensions, etc.
  }
  
  set physicalProperties(value: string | undefined) {
    if (value !== undefined) this.setProperty('physicalProperties', value);
  }

  get identifiers(): string | undefined {
    return this.getProperty<string>('identifiers'); // UID, tray ID, consumable ID
  }
  
  set identifiers(value: string | undefined) {
    if (value !== undefined) this.setProperty('identifiers', value);
  }

  get interpretationVersion(): string | undefined {
    return this.getProperty<string>('interpretationVersion'); // Version of interpretation logic
  }
  
  set interpretationVersion(value: string | undefined) {
    if (value !== undefined) this.setProperty('interpretationVersion', value);
  }

  get interpretationTime(): number | undefined {
    return this.getProperty<number>('interpretationTime');
  }
  
  set interpretationTime(value: number | undefined) {
    if (value !== undefined) this.setProperty('interpretationTime', value);
  }

  get cacheTimestamp(): Date {
    return this.getProperty<Date>('cacheTimestamp') ?? new Date();
  }
  
  set cacheTimestamp(value: Date) {
    this.setProperty('cacheTimestamp', value);
  }

  get cacheTtlMinutes(): number {
    return this.getProperty<number>('cacheTtlMinutes') ?? 15; // Shorter TTL for complex interpretations
  }
  
  set cacheTtlMinutes(value: number) {
    this.setProperty('cacheTtlMinutes', value);
  }

  isExpired(): boolean {
    const expiryTime = new Date(this.cacheTimestamp.getTime() + this.cacheTtlMinutes * 60 * 1000);
    return new Date() > expiryTime;
  }
}

/**
 * Cache management implementation for ephemeral entities
 */
export class CacheEntryImpl<T extends Entity> implements CacheEntry<T> {
  constructor(
    public readonly entity: T,
    public readonly timestamp: Date = new Date(),
    public readonly ttlMinutes: number
  ) {}

  isExpired(): boolean {
    const expiryTime = new Date(this.timestamp.getTime() + this.ttlMinutes * 60 * 1000);
    return new Date() > expiryTime;
  }

  getRemainingTtl(): number {
    const expiryTime = new Date(this.timestamp.getTime() + this.ttlMinutes * 60 * 1000);
    const now = new Date();
    return now < expiryTime 
      ? Math.floor((expiryTime.getTime() - now.getTime()) / (60 * 1000))
      : 0;
  }
}